### PR TITLE
feat(autoupdates): install updates without confirmation in the background by default COMPASS-7616

### PR DESCRIPTION
**Pull request by addaleax**
…

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass/src/main/auto-update-manager.ts [hunk 1]
```diff
@@ -155,67 +168,125 @@ const checkForUpdates: StateUpdateAction = async function checkForUpdates(
...
-          this: { maybeInterrupt(): void | never },
-          updateManager: typeof CompassAutoUpdateManager,
-          ...extraArgs: any[]
-        ) => void | Promise<void>
-      >
-    >
-  >
+const STATE_UPDATE: Record<
+  AutoUpdateManagerState,
+  AutoUpdateManagerStateDefinition
```
Comment (addaleax) [inline packages/compass/src/main/auto-update-manager.ts:194]: (heads up that Github hides the diff for this file by default)

This flips around the logic a bit; instead of defining specific actions to be taken when going from A to B (and the allowed set of state transitions being implied by the presence of an explicitly defined action), this defines an action to be performed when *entering* state B, and a valid set of states to transition to.

I hope this is okay (@gribnoysup you wrote this, so I'd like to check with you), but I'll have to say, I kind of had to do this in order for me to really feel like I covered the right set of changes here 😅 I think the fact that we had previously factored out functions like `manualCheck`, `checkForUpdates`, `disableAutoUpdates` so that we were able to use them for multiple [from, to] pairs was a sign to me that this would make sense, obviously we could keep following that pattern if we really felt like we want to keep it.

File: packages/compass-home/src/components/home.tsx [hunk 2]
```diff
@@ -315,6 +331,75 @@ function Home({
...
+      ipcRenderer?.removeListener(
+        'autoupdate:update-download-failed',
+        onAutoupdateFailed
+      );
+      ipcRenderer?.removeListener(
+        'autoupdate:update-download-success',
+        onAutoupdateSucess
+      );
+    };
+  }, []);
```
[the remark this answers is not in the document]
Comment (gribnoysup) [reply packages/compass-home/src/components/home.tsx:401]: I pinged you in tech design, seems like we haven't expanded there, but in the initial draft I added a sentence that it would be nice to abstract this behind an interface, yeah. I'm not sure about making toast specifically part of the interface, but some way of getting answers back from renderer  in main that does all the wiring of events on its own would be nice I think

[the remark this answers is not in the document]
Comment (addaleax) [reply packages/compass-home/src/components/home.tsx:401]: Yeah ... fair point, I was thinking about toasts specifically but even just better ways of doing these "reverse" calls would be nice.

I opened https://jira.mongodb.org/browse/COMPASS-7674, let's maybe first figure out what our solution should look like here before we implement it

File: packages/compass/src/main/auto-update-manager.ts [hunk 3]
```diff
@@ -583,10 +640,21 @@ class CompassAutoUpdateManager {
...
 
+    ipcMain?.handle(
+      'autoupdate:update-download-restart-confirmed',
+      this.handleIpcUpdateDownloadRestartConfirmed
+    );
+
+    ipcMain?.handle(
+      'autoupdate:update-download-restart-dismissed',
+      this.handleIpcUpdateDownloadRestartDismissed
+    );
```
[the remark this answers is not in the document]
Comment (addaleax) [reply packages/compass/src/main/auto-update-manager.ts:654]: https://jira.mongodb.org/browse/COMPASS-7673

File: packages/compass/src/main/auto-update-manager.ts (hunk 1 above)
Comment (gribnoysup) [reply packages/compass/src/main/auto-update-manager.ts:194]: Yeah, no strong preference here on how the state transitions are done. I did like the explicitness of the old state transition set up, but I also agree that it was hard to follow and make sure that there are no missing transitions. I like the change 🙂

**Review by gribnoysup — APPROVED** [review]

File: packages/compass-home/src/components/home.tsx [hunk 4]
```diff
@@ -315,6 +331,75 @@ function Home({
...
+  useEffect(() => {
+    function onAutoupdateStarted() {
+      openToast('update-download', {
+        variant: 'progress',
+        title: 'Compass update is in progress',
+      });
+    }
+    function onAutoupdateFailed() {
+      closeToast('update-download');
+      openToast('update-download', {
```
Comment (gribnoysup) [inline packages/compass-home/src/components/home.tsx:343]: Oh, one small thing that I forgot to mention, it's okay if you want to close it first, but technically `openToast` works closer to `updateToast` (we just copied leafygreen name for the method), so if all your toasts have the same id, they should just update when this is called, no need to close the previous ones first
